src/Psl/Regex/every_match.php

```diff
@@ -0,0 +1,50 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\Regex;
+
+use Psl\Collection\VectorInterface;
+use Psl\Type;
+
+use function preg_match_all;
+
+/**
+ * Determine if $subject matches the given $pattern and return every matches.
+ *
+ * @template T of array
+ *
+ * @param non-empty-string $pattern             The pattern to match against.
+ * @param Type\TypeInterface<T> $capture_groups What shape does a single set of matching items have?
+ *
+ * @return VectorInterface<T>|null
+ *
+ * @throws Exception\RuntimeException        If an internal error accord.
+ * @throws Exception\InvalidPatternException If $pattern is invalid.
+ */
+function every_match(
+    string $subject,
+    string $pattern,
+    Type\TypeInterface $capture_groups,
+    int $offset = 0
+): ?VectorInterface {
```

Comment:
I'm still not sure if this is the best way to do this ( `$capture_groups` ), the reason we have this argument is because we want to provide type safety, but is there another way we can provide type safety ( as well as runtime safety ) without having this argument?

i think we should look at other languages and how they implement this ( beside hack, because regex there is part of the type checker, and there's a special type for it ).

for example, Java has [`Matcher`](https://docs.oracle.com/javase/7/docs/api/java/util/regex/Matcher.html), which doesn't use generics, but provides type safety.

Reply:
There is no "easy way", since you'll need to parse the regex in order to detect the different capture groups in PHP.
I've seen this project: https://github.com/BackEndTea/Regexer - but we don't want to pull in dependencies and it might be a bit heavy during runtime as well.

Reply:
I've thought about it again and see 2 options:

1. Either you can do it like this. The advantage is that you can specify your own shape and cast the types to e.g. integers where it makes sense. The downside is that you manually need to declare it, even though the groups could be known from the regex.
2. Move the type part completely to the psalm plugin, meaning we'll have to write a custom regex leader that searched for capture groups and then calculates the returned type based on that information. Advantage is that type-safe group support will be inside the language and only be calculated during static analysis. The downside is that we'll need to write a custom regex parser, it only works with psalm and you cannot cast types from the function signature. An additional benefit could be that you can parse all regexes and look for regex errors from psalm for all regex functions if done right. however, thid whole feature looks like a completely decoupled regex psalm plugin to me (in which you could use the regex parser library shared above).

Reply:
parsing regex is the last things we really want 😛 unless we can find a parser that we can declare as an optional dependency and add note about it in the documentation so people can install it to get better static analysis for regex related things.